core/state/pruner/pruner.go

```diff
@@ -261,6 +261,8 @@ func (p *Pruner) Prune(root common.Hash) error {
 		}
 		// Use the bottom-most diff layer as the target
 		root = layers[len(layers)-1].Root()
+	} else if p.snaptree.Snapshot(root) == nil {
+		p.snaptree.Rebuild(root, false)
```

Comment:
This seems like it will work.. however, I think that a better alternative would be to skip creating the snaptree in NewPruner and passing the right root to snapshot.New which wouldn't require any change in snapshot.

For nitro - I think we'll need to support multiple roots (so all will be cleaned unless in any of these roots), which should probably be done with multiple snaptrees, each with a different root, each updating the same stateBloom.

At least in a validators we'd like to keep state for: something recent + latest validated + latest accepted onchain.

Reply:
I've added support for multiple roots, but I found that having multiple snaptrees was way too expensive, and would completely destroy any existing snapshots. Instead, I've added support for generating the pruning bloom filters without a snapshot :)